### ISSUE TITLE
ESLint should adhere to ignore rules that wrap import() and import.meta

### ISSUE THREAD

**Issue by klebba** [eslint/eslint#11514]
**The version of ESLint you are using.**

Latest (5.x.x at time of writing)

**The problem you want to solve.**
```
// eslint-disable-next-line
const module = import('./foo.js');
```

This line should be skipped rather than blow up with a parsing error.

**Your take on the correct solution to problem.**

ESLint could instead analyze my code for exclusion rules and then parse what remains. This would be the expected ignore behavior IMO, and not require a policy debate.

It seems totally reasonable to adopt only Stage 4 features, but on balance it seems completely unreasonable to not actually ignore this line. ESLint should consider a less pure solution that strikes a balance here. Folding in a transpile step via `babel-eslint` to make sure this ignore happens is not a good option. Ultimately I don't care if this line is skipped, but to have all the proximal code become collateral damage is a bummer.

Thanks in advance for considering!

Related:
* https://github.com/eslint/eslint/issues/11486

**Are you willing to submit a pull request to implement this change?**

I'm not currently informed enough to know the level of effort here, so I can't commit to making the change yet.

**Comment by not-an-aardvark** [comment]
Hi, thanks for creating an issue.

However, I'm not sure this is feasible, for a few reasons:

* Without parsing the file first, it's nontrivial to figure out what is actually a comment. For example, if someone's code is:

    ```js
    const disableMarker = "// eslint-disable-line";
    ```

    Presumably, they wouldn't want that line to be disabled just because they to be using the string `eslint-disable-line` in their application logic. (To give a contrived example, someone might have the string `eslint-disable-line` in their code if they were building tooling that works alongside ESLint and automatically disables certain warnings.)

* People sometimes use `eslint-disable-line` on lines that are syntactically necessary. For example, if someone's code is:

    ```js
    function getResult() {
        return import("foo").then(result => { // eslint-disable-line
            return result.doSomething();
        });
    }
    ```

    If ESLint entirely ignored the second line when parsing, it would end up trying to parse the following, which would cause an unrelated parsing error due to the unmatched parentheses on line 3:

    ```js
    function getResult() {
            return result.doSomething();
        });
    }
    ```

**Comment by not-an-aardvark** [comment]
Requiring both `/* eslint-disable */` and `/* eslint-disable */` would help with the second issue from https://github.com/eslint/eslint/issues/11514#issuecomment-473415680 in some cases, but would not address the first issue. It also seems potentially confusing that `/

Currently, `/* eslint-disable */` means "ESLint should suppress any errors reported for this section of the code", not "ESLint should pretend this section of the code doesn't exist at all". (To illustrate the difference: If ESLint entirely ignored the line containing `import` in your example, then it would also start reporting errors about the variable `module` not being defined if the variable was used later.) So while I sympathize with the inconvenience of having ESLint fail on parsing errors, I'm not convinced that being able to use `// eslint-disable` to suppress parsing errors is a desirable goal in the first place. Unlike many other linters, ESLint does separate passes for parsing and linting (this makes it possible to support arbitrary custom parsers and rules), and so there is naturally a fundamental difference between parsing errors and other types of linting errors. I'm not sure having `eslint-disable` comments conflate the two is a good idea.

One could imagine having a separate type of directive that does what you're asking (i.e. it strips out sections of code before the code gets parsed). But it seems to me that it would almost always be better to use a different parser that handles the case you want, rather than adding directives to slice that case out of the code before linting.

**Comment by klebba** [comment]
> "ESLint should suppress any errors reported for this section of the code"

Yes, this is exactly what I would expect, but not the current behavior. I understand that a change here would be fundamental if the parser can't interpret this code. For that to happen ESLint would have to pick and choose what else to interpret outside of Stage 4, which I understand is not the policy. Nonetheless, the behavior here is weird, which is why I think...

> ESLint should consider a less pure solution that strikes a balance here.

Could ESLint learn to ignore certain Stage 3 syntax rather than choke? Perhaps having the parser interpret `import()` and `import.meta` would be more valuable for end users of the library. Otherwise my options are:

1. Isolate the offending code in separate files
2. Add another dependency to every project (`babel-eslint`) and update my ESLint configuration everywhere to use it
3. Accept that ESLint will choke on any files using dynamic imports or needing to introspect their path

**Comment by ilyavolodin** [comment]
I think you are only using one example (that's very simple, in comparison to other cases). How, for example, would you have ever disabled `for...of` before it became Stage 4. Or `async\await`? If we were to consider some solution for disabling (or rather, hiding) some syntax from the parser, it would need to be generic. But in a lot of cases, it's not really possible to comment out new syntax without also commenting out a large portion of the code.
That's the reason `babel-eslint` exists. It should be able to parse pretty much anything that's above Stage 0.

**Comment by klebba** [comment]
This is how Firefox messages a developer about unsupported import functionality:
![Screen Shot 2019-04-05 at 11 32 59 AM](https://user-images.githubusercontent.com/287884/55649004-a0714700-5796-11e9-9506-db069829f5a6.png)

> If we were to consider some solution for disabling (or rather, hiding) some syntax from the parser, it would need to be generic.

If we consider the value proposition between handling the case for `import()` versus `for...of` it's asymmetric is it not? Just to reiterate:

> Perhaps having the parser interpret import() and import.meta so that it can be ignored would be more valuable for end users of the library and allow the maintainers to stay true to the policy.

**Comment by platinumazure** [comment]
> This is how Firefox messages a developer about unsupported import functionality:

Firefox and ESLint behave similarly in this case. Firefox identifies the SyntaxError and does not run the script in question due to being unable to create a parse tree. Similarly, ESLint identifies the error during parsing and does not lint the script due to being unable to create a parse tree.

> > Perhaps having the parser interpret import() and import.meta so that it can be ignored would be more valuable for end users of the library and allow the maintainers to stay true to the policy.

In theory, this could be done, but it would drastically increase the maintenance burden. We would constantly have to look at Stage 3 (or even lower) features and decide if we want to have a compatibility layer that can handle those, and then remove the compatibility/make changes as language features hit stage 4. As an all-volunteer team, we simply don't have the resources to maintain a parser like that, so we have chosen to focus on stage 4 language features only.

We intentionally designed ESLint to be highly pluggable for exactly this reason. As has been stated earlier, this includes being able to replace the parser. And babel-eslint supports this feature. So, there is a path forward for you here.

I don't think we're going to re-evaluate our policy on this, outside of the unlikely circumstance where we have a large number of people willing to step up and help us with the maintenance burden. And even then, ESLint could probably use that extra help more effectively in other areas (e.g., documentation improvements/translation or implementing long-sought-after but difficult features such as parallel linting). We appreciate the constructive discussion here, but given that we are constrained by our resource limitations and given that you have another option, I doubt this is something we will, change at this time.

**Comment by klebba** [comment]
@platinumazure thanks for your thoughtful reply. Perhaps I should be more clear about my comments:

- Though the syntax is _unsupported_ in Firefox, the error response is still tailored for imports, indicating that Firefox still understands something specific about the `import()` statement (perhaps because it's _planned_ syntax)
- I'm still not advocating for a change in Stage 3 _support_ -- I'm advocating that ESLint _not choke_ when I use `import()` and `import.meta` syntax
- My request is to specifically handle _only these cases_ so that I can continue to use ESLint for the syntax that it does support. The cases I have identified here are finite; maintaining them should not be a perpetual burden (if anything the burden will be lifted at Stage 4). I recognize the pushback from the maintainers about supporting _all Stage 3 cases_ but this has not been and will not be what I think is `a less pure solution that strikes a balance here`. I am specifically not recommending that change.
- ES6 import support in all browsers and in NodeJS will change the way developers build software; it's not syntactical sugar. The syntax is so fundamental (and bound to the `import` keyword) that I believe it is reasonable for ESLint to gracefully handle it instead of bailing. My suggestion to use ignore statements only matches my expectations for what they do; they tell ESLint not to flag that line for an error.

Hopefully this helps clarify the suggestion. I fully recognize that it smells like a policy question because the parser would have to change, but that is not my intent. My message here is about developer ease of use; I have been working around this problem and using ESLint since the functionality landed in Chrome: https://developers.google.com/web/updates/2017/11/dynamic-import -- I have no intention of adopting another dependency to solve this problem, I already outlined my workarounds above.

Thanks for considering

**Comment by ilyavolodin** [comment]
@klebba Thanks for suggestion. I understand your use-case and the problems you are running into. And I agree that workaround is inconvenient, but I think you are not taking into consideration our position. This case is special to you, because you are using it in your code. But ESLint has 7 million downloads a week, as such we have a lot of users who use all kind of specific features of the language. For somebody who is heavily invested into private class fields, it would be a bigger problem then not supporting dynamic imports. Or top level awaits. We can't take one use-case and say that we need to fix it because vocal minority needs it. We either have to fix all unsupported features, and basically incorporate babel-eslint into core, or we have to draw the line in the sand at Stage 4 features and say that's what we are going to support and suggest babel-eslint for all other cases. We also can't make parser ignore some of the features, because in order to ignore them, it needs to know about them (and ESLint parser, Espree, is built on top of Acorn, which also have to know about them as well). ESLint was built to work only on valid JS code. We describe "valid" as code that contains only syntax that reached Stage 4. As such, dynamic imports are not considered valid, and we can't support them.

**Comment by klebba** [comment]
@ilyavolodin I understand the desire to beat the drum of dogma again and again to reject this suggestion, but this repetition is not necessary. Your points have been made in this thread already. If the maintainers position is that no exceptions can be made under any circumstances because of the slippery slope argument, that's your call. I hope I've been _crystal clear_ that I'm strictly seeking an exception to the rule for `import`.

**Comment by johanneswilm** [comment]
Is there still no suggested workaround for this?

**Comment by platinumazure** [comment]
@johanneswilm As has been noted multiple times in this issue, you can use the `babel-eslint` parser to have ESLint work with that syntax.
